osu.Game/Screens/Play/MasterGameplayClockContainer.cs

```diff
@@ -197,6 +207,51 @@ public void StopUsingBeatmapClock()
             addAdjustmentsToTrack();
         }
 
+        protected override void Update()
+        {
+            base.Update();
+            checkPlaybackValidity();
+        }
+
+        #region Clock validation (ensure things are running correctly for local gameplay)
+
+        private double elapsedGameplayClockTime;
+        private double? elapsedValidationTime;
+        private int playbackDiscrepancyCount;
+
+        private const int allowed_playback_discrepancies = 5;
+
+        private void checkPlaybackValidity()
+        {
+            if (GameplayClock.IsRunning)
+            {
+                elapsedGameplayClockTime += GameplayClock.ElapsedFrameTime;
+
+                elapsedValidationTime ??= elapsedGameplayClockTime;
+                elapsedValidationTime += GameplayClock.Rate * Time.Elapsed;
```

Comment:
How does this work? If `elapsedValidationTime` is null in a frame, doesn't this cause said frame to be double-counted?

I'd expect something closer to

```csharp
                if (elapsedValidationTime == null)
                    elapsedValidationTime = elapsedGameplayClockTime;
                else
                    elapsedValidationTime += GameplayClock.Rate * Time.Elapsed;
```

Reply:
Yeah, that looks better. Good catch.